projects/RabbitMQ.Client/client/impl/ConsumerDispatching/ConsumerDispatcherBase.cs

```diff
@@ -0,0 +1,72 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace RabbitMQ.Client.client.impl.ConsumerDispatching
+{
+    internal abstract class ConsumerDispatcherBase
+    {
+        private readonly Dictionary<string, IBasicConsumer> _consumers;
+
+        public IBasicConsumer DefaultConsumer { get; set; }
+
+        protected ConsumerDispatcherBase()
+        {
+            _consumers = new Dictionary<string, IBasicConsumer>();
+        }
+
+        protected void AddConsumer(IBasicConsumer consumer, string tag)
+        {
+            lock (_consumers)
+            {
+                _consumers[tag] = consumer;
+            }
+        }
+
+        protected IBasicConsumer GetConsumerOrDefault(string tag)
+        {
+            lock (_consumers)
+            {
+                return _consumers.TryGetValue(tag, out var consumer) ? consumer : GetDefaultConsumerOrThrow();
+            }
+        }
+
+        public IBasicConsumer GetAndRemoveConsumer(string tag)
+        {
+            lock (_consumers)
+            {
+#if NETSTANDARD
+                var consumer = _consumers[tag];
```

Comment:
This line may cause the same exception as #1013 if the tag isn't in the dictionary.

Reply:
Yes, but question is, can this ever happen? If the server sends a tag twice or a wrong one, then causing an exception seems fine to me. (GetConsumerOrDefault on the other hand does need to care as the protocol allows scenario where this could happen.)

Reply:
in #1012, the system has tried to get a consumer that wasn't in the dictionary and it caused chaos in the client.  It may have been related to a network glitch with timeouts when unsubscribing a consumer that caused the message to not be there, or to get sent in from the server multiple times. The dictionary exception kept reoccurring consistently and frequently.  I haven't tracked down the root cause though, or spent much time trying to duplicate, but I have had that exception a enough times in production that it got on my radar to fix.

I don't think we should throw an exception here since it isn't handled well further up the stack.  Logging the error seems fairly reasonable though, and may help track down any bugs.  Then we can ignore subsequent actions, or resort to some default predictable behaviour (like defaultConsumer).

This function looks like it gets called on shutdown, so handling the error gracefully is probably not a big problem.  I'm not familiar enough with the full code base though to comment on this aspect authoritatively.

Reply:
Thanks for the information. I overlooked that it happened on CancelOk.

Yes, for CancelOk there might be some overlapping in some strange scenarios possible. [Look at](https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/997/files#diff-9b50eae409f49c27dddd300e8103618ff1c951c03cb1624f54484c3dd36250bfL969).

If for whatever reason the server sends some other reply instead of CancelOk, the consumer will be removed here and  later again when the actual CancelOk comes. This is the reason in my PR I removed this double removing.

But what interests me more is the failure to handle the exception more gracefully, I'll take a look at it sometime. 

AFAIK on protocol error (e.g. getting a consumer tag that doesn't exist) should result in an exception, @michaelklishin is that right? Or how should this be handled?

Reply:
The active consumer state is present in channel state on both sides, so there is a race condition between what the client and server do. E.g. you can call a `basic.cancel` and immediately remove the tag, only to get a `basic.cancel-ok` with it later. Bot sides in this example do something that makes sense at first glance but the relative timing is problematic.

In other clients we remove the consumer after receiving a `basic.cancel-ok`. If we don't have a consumer but received a consumer operation frame for it, some clients log a warning. Gracefully doing nothing would also be acceptable to me.

Reply:
I've implemented it now in a ultra safe manner, in which the consumer is:
- found via the tag, if not
- DefaultConsumer is returned, if null then
- FallbackConsumer is returned, where the implementation will just log the call itself and continues.

Reply:
@LamarLugli Would you agree that #1012 / #1013 is still fixed with the current proposed implementation?